sys/vfs/vfs.c

```diff
@@ -0,0 +1,660 @@
+    }
+    _vfs_open_files[fd].pid = KERNEL_PID_UNDEF;
+}
+
+inline static int _init_fd(int fd, const vfs_file_ops_t *f_op, vfs_mount_t *mountp, int flags, void *private_data)
+{
+    fd = _allocate_fd(fd);
+    if (fd < 0) {
+        return fd;
+    }
+    vfs_file_t *filp = &_vfs_open_files[fd];
+    filp->mp = mountp;
+    filp->f_op = f_op;
+    filp->flags = flags;
+    filp->pos = 0;
+    filp->private_data = private_data;
+    return fd;
+}
+
+inline static int _allocate_mount(const char *mount_point)
+{
+    for (int md = 0; md < VFS_MAX_OPEN_FILES; ++md) {
+        if (_vfs_mounts[md].mount_point[0] == '\0') {
+            vfs_mount_t *mountp = &_vfs_mounts[md];
+            int res = vfs_normalize_path(mountp->mount_point, mount_point, VFS_MOUNT_POINT_LEN);
+            if (res < 0) {
+                /* something is wrong with the given mount point string */
+                return res;
+            }
+            int len = strnlen(mountp->mount_point, VFS_MOUNT_POINT_LEN);
+            if (len < 1) {
+                /* something is wrong with the given mount point string */
+                _free_mount(md);
+                return -EINVAL;
+            }
+            if (mountp->mount_point[len - 1] == '/') {
+                /* remove trailing slash */
+                mountp->mount_point[len - 1] = '\0';
+            }
+            ATOMIC_VALUE(_vfs_mounts[md].open_files) = 0;
+            _vfs_mounts[md].private_data = NULL;
+            _vfs_mounts[md].fs = NULL;
+            return md;
+        }
+    }
+    /* The _vfs_mounts array is full */
+    return -ENOMEM;
+}
+
+inline static void _free_mount(int md)
+{
+    _vfs_mounts[md].mount_point[0] = '\0';
+}
+
+inline static int _find_mount(const char *name, const char **rel_path)
+{
+    size_t longest_match = 0;
+    size_t name_len = strnlen(name, VFS_MOUNT_POINT_LEN - 1);
+    int md;
+    mutex_lock(&_mount_mutex);
+    /* naive search for the longest matching prefix */
+    for (int i = 0; i < VFS_MAX_MOUNTS; ++i) {
+        size_t len = strnlen(_vfs_mounts[i].mount_point, VFS_MOUNT_POINT_LEN - 1);
+        if (len < longest_match) {
+            /* Already found a longer prefix */
+            continue;
+        }
+        else if (len > name_len) {
+            /* path name is shorter than the mount point name */
+            continue;
+        }
+        else if (name[len] != '/' && name[len] != '\0') {
+            /* name does not have a directory separator where mount point name ends */
+            continue;
+        }
+        /* test whether mount_point is a prefix of name */
+        if (strncmp(name, _vfs_mounts[i].mount_point, len) == 0) {
+            longest_match = len;
+            md = i;
+        }
+    }
+    if (longest_match == 0) {
+        /* not found */
+        mutex_unlock(&_mount_mutex);
+        return -ENOENT;
+    }
+    /* Increment open files counter for this mount */
+    atomic_inc(&_vfs_mounts[md].open_files);
```

Comment:
Actually, this is on purpose as a kind of semaphore to avoid races between umount and open/unlink. There were a few missing decrements of open_files in vfs_unlink though that I have fixed now.